src/libcore/slice/mod.rs

```diff
@@ -2074,6 +2074,85 @@ impl<T> [T] {
              from_raw_parts_mut(mut_ptr.add(rest.len() - ts_len), ts_len))
         }
     }
+
+    /// Checks if the elements of this slice are sorted.
+    ///
+    /// That is, for each element `a` and its following element `b`, `a <= b` must hold. If the
+    /// slice yields exactly zero or one element, `true` is returned.
+    ///
+    /// Note that if `Self::Item` is only `PartialOrd`, but not `Ord`, the above definition
+    /// implies that this function returns `false` if any two consecutive items are not
+    /// comparable.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(is_sorted)]
+    /// let empty: [i32; 0] = [];
+    ///
+    /// assert!([1, 2, 2, 9].is_sorted());
+    /// assert!(![1, 3, 2, 4].is_sorted());
+    /// assert!([0].is_sorted());
+    /// assert!(empty.is_sorted());
+    /// assert!(![0.0, 1.0, std::f32::NAN].is_sorted());
+    /// ```
+    #[inline]
+    #[unstable(feature = "is_sorted", reason = "new API", issue = "53485")]
+    pub fn is_sorted(&self) -> bool
+    where
+        T: PartialOrd,
+    {
+        self.is_sorted_by(|a, b| a.partial_cmp(b))
+    }
+
+    /// Checks if the elements of this slice are sorted using the given comparator function.
+    ///
+    /// Instead of using `PartialOrd::partial_cmp`, this function uses the given `compare`
+    /// function to determine the ordering of two elements. Apart from that, it's equivalent to
+    /// `is_sorted`; see its documentation for more information.
+    #[unstable(feature = "is_sorted", reason = "new API", issue = "53485")]
+    pub fn is_sorted_by<F>(&self, mut compare: F) -> bool
+    where
+        F: FnMut(&T, &T) -> Option<Ordering>
+    {
+        let mut last = match self.first() {
+            Some(e) => e,
+            None => return true,
+        };
+
+        for curr in &self[1..] {
+            if compare(&last, &curr).map(|o| o == Ordering::Greater).unwrap_or(true) {
+                return false;
+            }
+            last = &curr;
+        }
```

Comment:
This way it could be interesting to use `chunk_exact` which allow llvm to do [better assumption for bounds checks](https://github.com/rust-lang/rust/issues/47115#issuecomment-354715511).

Reply:
Good call, thanks! So it's been changed from
```
let mut last = match self.first() {
    Some(e) => e,
    None => return true,
};

for curr in &self[1..] {
    if compare(&last, &curr).map(|o| o == Ordering::Greater).unwrap_or(true) {
        return false;
    }
    last = &curr;
}

true
```
to
```
let len = self.len();
    if len <= 1 {
        return true;
    }

for i in 1..len {
    if compare(&self[i - 1], &self[i]).map(|o| o == Ordering::Greater).unwrap_or(true) {
        return false;
    }
}

true
```

If LLVM is able to elide (almost) all bounds checks on both implementations, and one is preferred over the other, let me know!

Reply:
In the second example, the first `if` condition is redundant and can be removed. If `len <= 1`, the `for` loop won't run at all. That was my initial idea behind my comment -- to make the code shorter. But removing this first `if` actually makes the generated assembly worse (it generates at least one bounds check). So yes, it should stay in there, but preferably with a comment explaining why it's there.

Regarding the generated code:

- [The first version with `last` variable](https://rust.godbolt.org/z/8uBjLe)
- [The second version with indexing](https://rust.godbolt.org/z/ynJIF3)

The second version is shorter. Otherwise they look fairly identical to me. But maybe someone with a bit more knowledge of these things should comment ^_^

I prefer the second version (indexing).

Reply:
Why don't we just use [`slice::windows`](https://doc.rust-lang.org/std/primitive.slice.html#method.windows) ?

https://rust.godbolt.org/z/4h4d42

Reply:
Looks like `while let` generates a little less: https://rust.godbolt.org/z/W6xDfr

Reply:
So what "matters" is not really the total number of instructions but the ones executed for each element (each loop iteration). It's 6 for [the indexing version](https://rust.godbolt.org/z/ynJIF3) and 7 for [the `windows(2)` version](https://rust.godbolt.org/z/4h4d42).

But a few remarks:
- Just counting instructions is not really a reliable way to guess the speed. So by now, we would just have to measure it to say anything about performance.
- All versions are pretty close to the optimum. And I don't expect there to be a huge difference between the versions.
- For small primitive types, we will have a SIMD implementation in the future, so we don't really need to care about that. For bigger types, the time to compare those types will be a significant part of the runtime, so again, this code doesn't matter *that* much.

So, I'd say to just take the **most idiomatic solution for now, which IMO is `for v in s.windows(2)`**. We can still build a bunch of benchmarks and improve the code later if it's necessary.

Reply:
So [indexing](https://rust.godbolt.org/z/ynJIF3) and [while let with `slice::windows`](https://rust.godbolt.org/z/8tGa54) both look to be good options.

At `opt-level=1` (if that is something to consider), indexing generates a little less.

At this point, like @LukasKalbertodt mentioned someone with a little more knowledge can make the call. I like how concise `while let` is, but if `opt-level=1` is what we should go by then I also lean towards indexing.

Reply:
Going with what's idiomatic initially and letting performance improve over time through targeted refactorings sounds like a good path forwards to me :+1: